src/cookbooks/client_guidelines.md

```diff
@@ -0,0 +1,137 @@
+  - If feature gates are [not appropriate](#feature-gate-caveats) for your project, Firefox Preferences (AKA "__Prefs__") should be used.
+- [Firefox Extensions](#extensions) AKA "__Add-ons__".
+  - If the feature being tested should not land in the tree, or if it will ultimately ship as an extension, then an extension should be used.
+
+New features go through the standard Firefox review, testing, and deployment processes, and are then enabled experimentally in the field using [Normandy](https://github.com/mozilla/normandy).
+
+## Feature Gates
+
+A new Feature Gate library for Firefox Desktop is now available.
+
+__Each feature gate should represent a different experimental treatment__.
+
+### Feature Gate caveats
+
+The current Feature Gate library comes with a few caveats, and may not be appropriate for your situation:
+
+- Only JS is supported.
+- Always asynchronous.
+
+Future versions of the Feature Gate API will include C++/Rust support and a synchronous API.
+
+### Using the Feature Gate library
+
+Read [the documentation][feature-gate-docs] to get started.
+
+## Prefs
+
+Firefox Preferences (AKA "prefs") are commonly used to enable and disable features. However, prefs are more complex to implement correctly than [feature gates](#feature-gates).
+
+__Each pref should represent a different experimental treatment__.
+
+There are three types of Prefs:
+
+1. Built-in prefs - shipped with Firefox, in `firefox.js`.
+2. `user branch` - set by the user, overriding built-in prefs.
+3. `default branch` - Overrides both built-in and `user branch` prefs. Only persists until the browser session ends, next restart will revert to either built-in or `user branch` (if set).
+
+  [Normandy](https://github.com/mozilla/normandy) supports overriding both the `user` and `default` branches, although the latter is preferred as it does not permanently override user settings. `default` branch prefs are simple to reset since they do not persist past a restart.
+
+   __In order for features to be activated experimentally, these rules must be observed__:
+
+- The feature must not start up before `final-ui-startup` is observed.
+- It must be possible to enable/disable the feature at runtime, via a pref change.
+- Use `default branch`, not `user branch` prefs.
+  - For instance, never use `Services.prefs.prefHasUserValue()`.
+- Prefs should be set by default in `firefox.js`
+
+If your feature cannot abide by one or more of these rules (for instance, it needs to run at startup and/or cannot be toggled at runtime) then experimental preferences can be set on the `user branch`.
+
+## Extensions
+
+Determining whether an extension is appropriate for an experiment will be decided on a case-by-case basis. As in the “Feature Gates” section above, landing features directly in-tree is preferred to using an extension for experiments.
+
+__Each extension can represent a different experimental treatment (preferred), or the extension can choose the branch internally__.
```

Comment:
This is the way upcoming search experiments are doing it. I think it's better because:

- these are just normal extensions
  - maybe with a few touches like telemetry and being able to end the study, but maybe not even that!
- We should be doing a lot more automated testing, which can test all of this things. There's no reason we can't mandate testing basic function(s) specific to the extension for startup/shutdown/(un)install,private browsing mode, etc.
- I'm not sure it's more burden for QA, I'd be happy to do it along with them to see if the workflow is particularly nasty.

Reply:
> these are just normal extensions

We can't guarantee that is the case always or even most of the time. For example I have never developed an experiment that was a pure WE.

> We should be doing a lot more automated testing

I agree, but how do we assure QA and ourselves that our tests cover all of these scenarios adequately and can be trusted over manual testing? Until this is standardized, I only see this increasing QA's and the developer's workload.

> I'd be happy to do it along with them to see if the workflow is particularly nasty

Yes I think it'd be a good idea to try this approach yourself to see its net effect before advocating it broadly. Maybe it's my own ShieldUtils bias, but I am having trouble seeing this as easier without a lot of automation and automated testing work. Maybe it's an initial investment that pays off over time?

Can you link to the upcoming search experiments' source code? I'm curious how they are doing it.

Reply:
They aren't quite ready on the upcoming experiment, let's take that 1:1 if you don't mind.